### PR TITLE
feat(jobs): add jobs for linkedlist rbegin/rend

### PR DESCRIPTION
**Pull request by danblitzhou**
- added jobs for `aws_linked_list_rbegin` and  `aws_linked_list_rend`
- added `ensure_linked_list_is_allocated_nd` in helper library; this function is useful for verifying linked list functions that only concern the front item and the back item (other than head and tail).

### CODE REVIEW

**Comment by agurfinkel** [comment]
close, but lots of redundancies that raises some questions. Let's clean it up before accepting.

File: seahorn/jobs/linked_list_rend/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+  aws_linked_list_rend_harness.c
+  ${SEA_LIB}/proof_allocators.c
+  ${SEA_LIB}/linked_list_helper.c)
+target_compile_definitions(linked_list_rend
+  # We will use our own function for is_valid check hence AWS_DEEP_CHECK is off
+  PUBLIC AWS_DEEP_CHECKS=0
+  MAX_BUFFER_SIZE=${MAX_BUFFER_SIZE}
+  MAX_LINKED_LIST_ITEM_ALLOCATION=${MAX_LINKED_LIST_ITEM_ALLOCATION_SIZE})
+sea_attach_bc(linked_list_rend)
+configure_file(sea.yaml sea.yaml @ONLY)
```
Comment (danblitzhou) [inline seahorn/jobs/linked_list_rend/CMakeLists.txt:13]: no job sea.yaml, will remove

Comment (agurfinkel) [reply seahorn/jobs/linked_list_rend/CMakeLists.txt:13]: also max size

**Review by agurfinkel — CHANGES_REQUESTED** [review]

File: seahorn/jobs/linked_list_rbegin/aws_linked_list_rbegin_harness.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    size_t length;
+    ensure_linked_list_is_allocated_nd(&list, MAX_LINKED_LIST_ITEM_ALLOCATION, &length);
+    if (length > 2) {
+        // hack to make sure tail.prev has valid prev node and next node
+        struct aws_linked_list_node *new_last = malloc(
+            sizeof(struct aws_linked_list_node));
+        aws_linked_list_push_back(&list, new_last);
+    }
+    /* Assume the preconditions */
+    assume(aws_linked_list_is_valid(&list));
```
[the remark this answers is not in the document]
Comment (danblitzhou) [reply seahorn/jobs/linked_list_rbegin/aws_linked_list_rbegin_harness.c:17]: It is if we keep `AWS_DEEP_CHECKS=0`; this essentially checks `list && list->head.next && list->head.prev == NULL && list->tail.prev && list->tail.next == NULL`; I believe I can make this compatible with the additional length > 2 models by adding `assume(list->head.next)` and `assume(list->tail.prev)`

File: seahorn/lib/linked_list_helper.c [hunk 3]
```diff
@@ -39,6 +38,37 @@ void ensure_linked_list_is_allocated(struct aws_linked_list *const list,
   }
 }
 
+void ensure_linked_list_is_allocated_nd(struct aws_linked_list *const list,
+                                            size_t max_length, size_t *length) {
```
[the remark this answers is not in the document]
Comment (danblitzhou) [reply seahorn/lib/linked_list_helper.c:42]: does `ensure_linked_list_is_initialized_nd` sound more reasonable?

File: seahorn/jobs/linked_list_rbegin/aws_linked_list_rbegin_harness.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (agurfinkel) [reply seahorn/jobs/linked_list_rbegin/aws_linked_list_rbegin_harness.c:17]: maybe better to have our version to have full control of the pf

File: seahorn/lib/linked_list_helper.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (agurfinkel) [reply seahorn/lib/linked_list_helper.c:42]: not really. 

there are 3 different functions like this that we seem to need. The names should reflect that. Check Sidharth PR that is still open. 

I think we are using linked_front for forward linked list, linked_back for back linked.

**Comment by danblitzhou** [comment]
addressed review comments:
- broke `ensure_linked_list_is_allocated_nd` into three function for different use case (specifically for length > 2)
- enabled `sea_aws_linked_list_is_valid` to check linked list shallowly, behaviour can be switched using macro constant `SEA_DEEP_CHECKS` similar to `AWS_DEEP_CHECKS`
- other renaming and small tweaks

File: seahorn/jobs/linked_list_rend/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# add head and tail to list size.
+MATH(EXPR UNROLL_BOUND "${MAX_LINKED_LIST_ITEM_ALLOCATION_SIZE} + 2")
+add_executable(linked_list_rend
+  aws_linked_list_rend_harness.c
+  ${SEA_LIB}/proof_allocators.c
+  ${SEA_LIB}/linked_list_helper.c)
+target_compile_definitions(linked_list_rend
+  # We will use our own function for is_valid check hence AWS_DEEP_CHECK is off
```
Comment (danblitzhou) [inline seahorn/jobs/linked_list_rend/CMakeLists.txt:8]: promise is met!

File: seahorn/jobs/linked_list_rbegin/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# add head and tail to list size.
+MATH(EXPR UNROLL_BOUND "${MAX_LINKED_LIST_ITEM_ALLOCATION_SIZE} + 2")
```
Comment (agurfinkel) [inline seahorn/jobs/linked_list_rbegin/CMakeLists.txt:2]: not needed

File: seahorn/jobs/linked_list_rbegin/aws_linked_list_rbegin_harness.c [hunk 6]
```diff
@@ -0,0 +1,40 @@
+#include <seahorn/seahorn.h>
+#include <aws/common/linked_list.h>
+#include <linked_list_helper.h>
+
+int main(void) {
+    /* data structure */
+    struct aws_linked_list list;
+    size_t length;
+    sea_nd_init_aws_linked_list_from_back(&list, &length);
+    // if (length > 2) {
```
Comment (agurfinkel) [inline seahorn/jobs/linked_list_rbegin/aws_linked_list_rbegin_harness.c:10]: remove

File: seahorn/jobs/linked_list_rend/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,12 @@
+# add head and tail to list size.
+MATH(EXPR UNROLL_BOUND "${MAX_LINKED_LIST_ITEM_ALLOCATION_SIZE} + 2")
```
Comment (agurfinkel) [inline seahorn/jobs/linked_list_rend/CMakeLists.txt:2]: not needed

**Review by agurfinkel — CHANGES_REQUESTED** [review]
almost there

**Comment by danblitzhou** [comment]
last review comments addressed

**Review by agurfinkel — APPROVED** [review]

**Pull request merged**